In [1]:
%config Completer.use_jedi = False 

import numpy as np
import pandas as pd

# 1 Datasets Preliminary Inspection #

Two datasets will compound our DataFrame:

**1) Weather Station Dataset:** With actuals values of the target variables(wind, wind gust, wind directions) measured on a weather station at the site.

**2) Historical Forecast Dataset:** With historical forecasts made in the past that we would use as features to train our improved forecast model.

## 1.1 Weather Station Dataset Inspection

Let s do a preliminary inspection of the weather station data before creating and manipulationg the work dataframe.


In [5]:
# Exploring Data Stations Data Set

df = pd.read_csv('Data/StationsRecords/DH-6001.csv', sep = ';', header = None)

In [6]:
df

0        1          2     3       4                    5    6   7   \
0      6001 -5.59883  36.013985  32.0  TARIFA  2019-02-16T09:00:00  0.0 NaN   
1      6001 -5.59883  36.013985  32.0  TARIFA  2019-02-16T10:00:00  0.0 NaN   
2      6001 -5.59883  36.013985  32.0  TARIFA  2019-02-16T11:00:00  0.0 NaN   
3      6001 -5.59883  36.013985  32.0  TARIFA  2019-02-16T12:00:00  0.0 NaN   
4      6001 -5.59883  36.013985  32.0  TARIFA  2019-02-16T13:00:00  0.0 NaN   
...     ...      ...        ...   ...     ...                  ...  ...  ..   
19105  6001 -5.59883  36.013985  32.0  TARIFA  2021-05-22T03:00:00  0.0 NaN   
19106  6001 -5.59883  36.013985  32.0  TARIFA  2021-05-22T04:00:00  0.0 NaN   
19107  6001 -5.59883  36.013985  32.0  TARIFA  2021-05-22T05:00:00  0.0 NaN   
19108  6001 -5.59883  36.013985  32.0  TARIFA  2021-05-22T06:00:00  0.0 NaN   
19109  6001 -5.59883  36.013985  32.0  TARIFA  2021-05-22T07:00:00  0.0 NaN   

       8   9   ...    29    30    31    32  33  34  35  36     37  38  
0     NaN NaN  ...  13.8  12.8  13.5  13.8 NaN NaN NaN NaN  412.0 NaN  
1     NaN NaN  ...  13.9  12.4  13.7  14.4 NaN NaN NaN NaN  388.0 NaN  
2     NaN NaN  ...  14.5  12.5  13.8  14.7 NaN NaN NaN NaN  408.0 NaN  
3     NaN NaN  ...  14.5  12.7  14.2  14.7 NaN NaN NaN NaN  401.0 NaN  
4     NaN NaN  ...  14.3  12.5  14.1  14.4 NaN NaN NaN NaN  421.0 NaN  
...    ..  ..  ...   ...   ...   ...   ...  ..  ..  ..  ..    ...  ..  
19105 NaN NaN  ...  16.7  15.9  16.6  17.1 NaN NaN NaN NaN  127.0 NaN  
19106 NaN NaN  ...  16.3  15.7  16.3  16.7 NaN NaN NaN NaN  110.0 NaN  
19107 NaN NaN  ...  17.0  16.8  16.2  17.0 NaN NaN NaN NaN   84.0 NaN  
19108 NaN NaN  ...  17.2  16.9  16.8  17.5 NaN NaN NaN NaN  105.0 NaN  
19109 NaN NaN  ...  17.8  16.8  17.2  17.8 NaN NaN NaN NaN   86.0 NaN  

[19110 rows x 39 columns]

In [7]:
df.isnull().sum()

0         0
1         0
2         0
3         0
4         0
5         0
6         0
7     19110
8     19110
9     19110
10       96
11       63
12    19110
13    19110
14       55
15    19110
16       96
17    19110
18    19110
19    19110
20    19110
21    19110
22       55
23    19110
24     3199
25     3203
26    19110
27    19110
28    19110
29       54
30       64
31       54
32       54
33    19110
34    19110
35    19110
36    19110
37        0
38    19110
dtype: int64

In [8]:
# Importing just data-filled columns.
df = pd.read_csv('Data/StationsRecords/DH-4554X.csv', sep = ';',
                 header = None,
                 usecols=[0,1,2,3,4,5,6,10,11,14,16,22,29,31,32],
                )

In [9]:
# Columns descrpition given within the documentation delivered with the csv in the data folder.

colnames = {0:'StationId',
           1:'Longitud',
           2:'Latitude',
           3:'Altutude',
            4:'StationName',
            5:'DateTime',
            6:'Precipitation',
            10:'WindMax',
            11:'Wind',
            14:'WindDirection',
            16:'WindDirectionMax',
            22:'RelHumidity',
            29:'AirTemperature',
            31:'AirMinTemp',
            32:'AirMaxTemp'}

df.rename(columns=colnames, inplace=True)

In [10]:
df.head()

StationId  Longitud   Latitude  Altutude        StationName  \
0     4554X  -7.08361  37.218334      15.0  CARTAYA   PEMARES   
1     4554X  -7.08361  37.218334      15.0  CARTAYA   PEMARES   
2     4554X  -7.08361  37.218334      15.0  CARTAYA   PEMARES   
3     4554X  -7.08361  37.218334      15.0  CARTAYA   PEMARES   
4     4554X  -7.08361  37.218334      15.0  CARTAYA   PEMARES   

              DateTime  Precipitation  WindMax  Wind  WindDirection  \
0  2019-02-16T09:00:00            0.0      2.8   1.5           92.0   
1  2019-02-16T10:00:00            0.0      4.3   1.9          110.0   
2  2019-02-16T11:00:00            0.0      4.8   2.1          100.0   
3  2019-02-16T12:00:00            0.0      5.2   2.9          184.0   
4  2019-02-16T13:00:00            0.0      5.3   2.0          209.0   

   WindDirectionMax  RelHumidity  AirTemperature  AirMinTemp  AirMaxTemp  
0             102.0         93.0             9.9         8.5         9.9  
1             117.0         82.0            12.1        10.3        12.1  
2              97.0         77.0            13.7        12.3        13.7  
3             190.0         66.0            16.1        14.1        16.1  
4             200.0         56.0            17.3        16.3        17.3

*Data from Stations is given in metric system, we are going to maintain that for all the variables but for the wind speed, that are going to be expressed in knots as is it widely used in nautical domains 1 m/s. = 1.9438.*

In [11]:
df.isna().sum()

StationId            0
Longitud             0
Latitude             0
Altutude             0
StationName          0
DateTime             0
Precipitation       14
WindMax             13
Wind                11
WindDirection        2
WindDirectionMax     0
RelHumidity         50
AirTemperature      50
AirMinTemp          50
AirMaxTemp          50
dtype: int64

In [12]:
df.describe()

Longitud      Latitude  Altutude  Precipitation       WindMax  \
count  1.930500e+04  1.930500e+04   19305.0   19291.000000  19292.000000   
mean  -7.083610e+00  3.721833e+01      15.0       0.044321      5.114659   
std    1.291445e-12  1.432491e-11       0.0       0.469344      2.645286   
min   -7.083610e+00  3.721833e+01      15.0       0.000000      0.000000   
25%   -7.083610e+00  3.721833e+01      15.0       0.000000      3.000000   
50%   -7.083610e+00  3.721833e+01      15.0       0.000000      4.700000   
75%   -7.083610e+00  3.721833e+01      15.0       0.000000      6.800000   
max   -7.083610e+00  3.721833e+01      15.0      18.200000     25.800000   

              Wind  WindDirection  WindDirectionMax   RelHumidity  \
count  19294.00000   19303.000000      19305.000000  19255.000000   
mean       2.44831     209.809201        209.048744     75.381927   
std        1.41586     110.732486        108.344691     17.747788   
min        0.00000       1.000000          0.000000     12.000000   
25%        1.40000     124.000000        127.000000     63.000000   
50%        2.20000     235.000000        235.000000     78.000000   
75%        3.30000     311.000000        305.000000     91.000000   
max        9.90000     360.000000        357.000000    100.000000   

       AirTemperature    AirMinTemp    AirMaxTemp  
count    19255.000000  19255.000000  19255.000000  
mean        16.974812     16.505214     17.440556  
std          6.092386      6.048396      6.119123  
min         -1.200000     -1.200000     -0.900000  
25%         12.700000     12.300000     13.200000  
50%         16.900000     16.500000     17.300000  
75%         21.000000     20.500000     21.500000  
max         41.000000     40.200000     41.900000

In [13]:
df.dtypes

StationId            object
Longitud            float64
Latitude            float64
Altutude            float64
StationName          object
DateTime             object
Precipitation       float64
WindMax             float64
Wind                float64
WindDirection       float64
WindDirectionMax    float64
RelHumidity         float64
AirTemperature      float64
AirMinTemp          float64
AirMaxTemp          float64
dtype: object

## 1.2 Historical Forecast Dataset Inspection

This  big dataset would require a big computing time if we open it completely. Then,  were going to open just few rows to inspect.


In [14]:
historicalForecast = pd.read_csv('Data/historicalForecast.csv', nrows = 1000)


In [15]:
pd.set_option('display.max_rows', 100)
historicalForecast.head(100)

forecast dt unixtime                forecast dt iso  slice dt unixtime  \
0             1507507200  2017-10-09 00:00:00 +0000 UTC         1507507200   
1             1507507200  2017-10-09 00:00:00 +0000 UTC         1507518000   
2             1507507200  2017-10-09 00:00:00 +0000 UTC         1507528800   
3             1507507200  2017-10-09 00:00:00 +0000 UTC         1507539600   
4             1507507200  2017-10-09 00:00:00 +0000 UTC         1507550400   
5             1507507200  2017-10-09 00:00:00 +0000 UTC         1507561200   
6             1507507200  2017-10-09 00:00:00 +0000 UTC         1507572000   
7             1507507200  2017-10-09 00:00:00 +0000 UTC         1507582800   
8             1507507200  2017-10-09 00:00:00 +0000 UTC         1507593600   
9             1507507200  2017-10-09 00:00:00 +0000 UTC         1507604400   
10            1507507200  2017-10-09 00:00:00 +0000 UTC         1507615200   
11            1507507200  2017-10-09 00:00:00 +0000 UTC         1507626000   
12            1507507200  2017-10-09 00:00:00 +0000 UTC         1507636800   
13            1507507200  2017-10-09 00:00:00 +0000 UTC         1507647600   
14            1507507200  2017-10-09 00:00:00 +0000 UTC         1507658400   
15            1507507200  2017-10-09 00:00:00 +0000 UTC         1507669200   
16            1507507200  2017-10-09 00:00:00 +0000 UTC         1507680000   
17            1507507200  2017-10-09 00:00:00 +0000 UTC         1507690800   
18            1507507200  2017-10-09 00:00:00 +0000 UTC         1507701600   
19            1507507200  2017-10-09 00:00:00 +0000 UTC         1507712400   
20            1507507200  2017-10-09 00:00:00 +0000 UTC         1507723200   
21            1507507200  2017-10-09 00:00:00 +0000 UTC         1507734000   
22            1507507200  2017-10-09 00:00:00 +0000 UTC         1507744800   
23            1507507200  2017-10-09 00:00:00 +0000 UTC         1507755600   
24            1507507200  2017-10-09 00:00:00 +0000 UTC         1507766400   
25            1507507200  2017-10-09 00:00:00 +0000 UTC         1507777200   
26            1507507200  2017-10-09 00:00:00 +0000 UTC         1507788000   
27            1507507200  2017-10-09 00:00:00 +0000 UTC         1507798800   
28            1507507200  2017-10-09 00:00:00 +0000 UTC         1507809600   
29            1507507200  2017-10-09 00:00:00 +0000 UTC         1507820400   
30            1507507200  2017-10-09 00:00:00 +0000 UTC         1507831200   
31            1507507200  2017-10-09 00:00:00 +0000 UTC         1507842000   
32            1507507200  2017-10-09 00:00:00 +0000 UTC         1507852800   
33            1507507200  2017-10-09 00:00:00 +0000 UTC         1507863600   
34            1507507200  2017-10-09 00:00:00 +0000 UTC         1507874400   
35            1507507200  2017-10-09 00:00:00 +0000 UTC         1507885200   
36            1507507200  2017-10-09 00:00:00 +0000 UTC         1507896000   
37            1507507200  2017-10-09 00:00:00 +0000 UTC         1507906800   
38            1507507200  2017-10-09 00:00:00 +0000 UTC         1507917600   
39            1507507200  2017-10-09 00:00:00 +0000 UTC         1507928400   
40            1507507200  2017-10-09 00:00:00 +0000 UTC         1507939200   
41            1507507200  2017-10-09 00:00:00 +0000 UTC         1507950000   
42            1507507200  2017-10-09 00:00:00 +0000 UTC         1507960800   
43            1507507200  2017-10-09 00:00:00 +0000 UTC         1507971600   
44            1507507200  2017-10-09 00:00:00 +0000 UTC         1507982400   
45            1507507200  2017-10-09 00:00:00 +0000 UTC         1507993200   
46            1507507200  2017-10-09 00:00:00 +0000 UTC         1508004000   
47            1507507200  2017-10-09 00:00:00 +0000 UTC         1508014800   
48            1507507200  2017-10-09 00:00:00 +0000 UTC         1508025600   
49            1507507200  2017-10-09 00:00:00 +0000 UTC         1508036400   
50            1507507200  

We have two kind of dates in each row: 

1) forecast dt(when the forecast was made)

2) slice dt (date for witch the forecast was made)

We re going to check how many slice dt are predicted in each forecast dt:

In [16]:
historicalForecast[historicalForecast['forecast dt unixtime']==1507528800]

forecast dt unixtime                forecast dt iso  slice dt unixtime  \
93             1507528800  2017-10-09 06:00:00 +0000 UTC         1507528800   
94             1507528800  2017-10-09 06:00:00 +0000 UTC         1507539600   
95             1507528800  2017-10-09 06:00:00 +0000 UTC         1507550400   
96             1507528800  2017-10-09 06:00:00 +0000 UTC         1507561200   
97             1507528800  2017-10-09 06:00:00 +0000 UTC         1507572000   
98             1507528800  2017-10-09 06:00:00 +0000 UTC         1507582800   
99             1507528800  2017-10-09 06:00:00 +0000 UTC         1507593600   
100            1507528800  2017-10-09 06:00:00 +0000 UTC         1507604400   
101            1507528800  2017-10-09 06:00:00 +0000 UTC         1507615200   
102            1507528800  2017-10-09 06:00:00 +0000 UTC         1507626000   
103            1507528800  2017-10-09 06:00:00 +0000 UTC         1507636800   
104            1507528800  2017-10-09 06:00:00 +0000 UTC         1507647600   
105            1507528800  2017-10-09 06:00:00 +0000 UTC         1507658400   
106            1507528800  2017-10-09 06:00:00 +0000 UTC         1507669200   
107            1507528800  2017-10-09 06:00:00 +0000 UTC         1507680000   
108            1507528800  2017-10-09 06:00:00 +0000 UTC         1507690800   
109            1507528800  2017-10-09 06:00:00 +0000 UTC         1507701600   
110            1507528800  2017-10-09 06:00:00 +0000 UTC         1507712400   
111            1507528800  2017-10-09 06:00:00 +0000 UTC         1507723200   
112            1507528800  2017-10-09 06:00:00 +0000 UTC         1507734000   
113            1507528800  2017-10-09 06:00:00 +0000 UTC         1507744800   
114            1507528800  2017-10-09 06:00:00 +0000 UTC         1507755600   
115            1507528800  2017-10-09 06:00:00 +0000 UTC         1507766400   
116            1507528800  2017-10-09 06:00:00 +0000 UTC         1507777200   
117            1507528800  2017-10-09 06:00:00 +0000 UTC         1507788000   
118            1507528800  2017-10-09 06:00:00 +0000 UTC         1507798800   
119            1507528800  2017-10-09 06:00:00 +0000 UTC         1507809600   
120            1507528800  2017-10-09 06:00:00 +0000 UTC         1507820400   
121            1507528800  2017-10-09 06:00:00 +0000 UTC         1507831200   
122            1507528800  2017-10-09 06:00:00 +0000 UTC         1507842000   
123            1507528800  2017-10-09 06:00:00 +0000 UTC         1507852800   
124            1507528800  2017-10-09 06:00:00 +0000 UTC         1507863600   
125            1507528800  2017-10-09 06:00:00 +0000 UTC         1507874400   
126            1507528800  2017-10-09 06:00:00 +0000 UTC         1507885200   
127            1507528800  2017-10-09 06:00:00 +0000 UTC         1507896000   
128            1507528800  2017-10-09 06:00:00 +0000 UTC         1507906800   
129            1507528800  2017-10-09 06:00:00 +0000 UTC         1507917600   
130            1507528800  2017-10-09 06:00:00 +0000 UTC         1507928400   
131            1507528800  2017-10-09 06:00:00 +0000 UTC         1507939200   
132            1507528800  2017-10-09 06:00:00 +0000 UTC         1507950000   
133            1507528800  2017-10-09 06:00:00 +0000 UTC         1507960800   
134            1507528800  2017-10-09 06:00:00 +0000 UTC         1507971600   
135            1507528800  2017-10-09 06:00:00 +0000 UTC         1507982400   
136            1507528800  2017-10-09 06:00:00 +0000 UTC         1507993200   
137            1507528800  2017-10-09 06:00:00 +0000 UTC         1508004000   
138            1507528800  2017-10-09 06:00:00 +0000 UTC         1508014800   
139            1507528800  2017-10-09 06:00:00 +0000 UTC         1508025600   
140            1507528800  2017-10-09 06:00:00 +0000 UTC         1508036400   
141            1507528800  2017-10-09 06:00:00 +0000 UTC         1508047200   
142            1507528800  2017-10-09 06:00:00 +0000 UT

... so for each forecast dt, we have 93 silce dt predictions, first of them given by 3 hours intervals and latest of them given by 12 h intervals.

Let see what is the frequency of forecast dt (how often a new forecast is provided):

In [17]:
historicalForecast['forecast dt iso'].value_counts()

2017-10-09 00:00:00 +0000 UTC    93
2017-10-09 06:00:00 +0000 UTC    93
2017-10-09 12:00:00 +0000 UTC    93
2017-10-09 18:00:00 +0000 UTC    93
2017-10-10 00:00:00 +0000 UTC    93
2017-10-10 06:00:00 +0000 UTC    93
2017-10-10 12:00:00 +0000 UTC    93
2017-10-10 18:00:00 +0000 UTC    93
2017-10-11 00:00:00 +0000 UTC    93
2017-10-11 06:00:00 +0000 UTC    93
2017-10-11 12:00:00 +0000 UTC    70
Name: forecast dt iso, dtype: int64

 ... so we can see there are 4 predictions daily (by 00.00, 06.00, 12.00, 18.00). That means for an specific datetime there will 93 different predictions, that will be updated each 6 hours. Obviously, the more recent the forecast the more accurate it will be. 



# 2 DataFrame Preparation & Manipulation

## 2.1 Weather Stations DataFrame Manipulation

We are going to fetch just wind meassurements at Weather Stations (Wind, Wind Gust, Wind direction) that will be the target variables to be determined with the features from the historicalweather dataset. The time granularity at historical weather dataset is 3h so we are going to use the same in the actual wind dataset. 

***In the first iteration we are going to collect just actual wind data(target) and just for the target location (tarifa, SPAIN, lon = -5.59883, lat=36.013985  ), but might consider collect data from other stations in further iterations that can help to improve the predictions.***

In [18]:
weatherStation_tarifa = pd.read_csv('Data/StationsRecords/DH-6001.csv', sep = ';',
                 header = None,
                 usecols=[5, 11,10,14]
                )

In [19]:
colnames = {5:'DateTime',
            11:'Tarifa_Wind',
            10:'Tarifa_WindGust',
            14:'Tarifa_WindDirection',
            }

weatherStation_tarifa.rename(columns=colnames, inplace=True)

In [20]:
weatherStation_tarifa.head()

DateTime  Tarifa_WindGust  Tarifa_Wind  Tarifa_WindDirection
0  2019-02-16T09:00:00             18.9          9.0                  72.0
1  2019-02-16T10:00:00             17.8          8.5                  78.0
2  2019-02-16T11:00:00             17.5          9.7                  89.0
3  2019-02-16T12:00:00             17.0         10.4                  86.0
4  2019-02-16T13:00:00             18.7         11.1                  83.0

In [21]:
weatherStation_tarifa.isna().sum()

DateTime                 0
Tarifa_WindGust         96
Tarifa_Wind             63
Tarifa_WindDirection    55
dtype: int64

**Things to be taken ino account:**

- **Granularity** : We want to adapt the granularity of Actuals(1h) to Historical Forecast (3h). For that purpose we are going to use the mean values of wind, gust and direction every three hours.

*A brief analysis between addapting the granuarity of the actuals to the granuarlity of the forecasts or viceversa. If we would have decide to addapt the granularity of forecasts (3h) to the actuals one(1h), we would have to interpolate values, then creating non-real values. Conversely, if we addapt actuals (1h) to forecast (3h), we would do a mean of 3 real actual meassurements, what is a real value.*


- **Direction of the wind:** Wind direction is given in degrees so it has to be taken into account that 360º and 0º are the same angle (N). A transformation need to be done to allow the model to understand that they are similar values.  

The strategy will be:

**STEP 1** Addapt granularity ot 3 h.

**STEP 2** Wind linear component to allow Machine Learning Model to better undestand the wind direciton (359º similar to 1º)

**STEP 3** Wind Speed in knots, that suits better in nautical environment. 





 - **STEP 1 : GRANULARITY**   Mean values for Wind and Wind Gust every 3 hours (granularity)

In [22]:
%%time

for i in weatherStation_tarifa.index:
    
    if int(weatherStation_tarifa['DateTime'][i][-8:-6])%3==0:
        weatherStation_tarifa.loc[i,'Tarifa_Wind'] = weatherStation_tarifa.loc[i:i+2,'Tarifa_Wind'].mean()
        weatherStation_tarifa.loc[i,'Tarifa_WindGust'] = weatherStation_tarifa.loc[i:i+2,'Tarifa_WindGust'].mean()
    
    else:
        weatherStation_tarifa.drop(i, axis=0, inplace=True)

weatherStation_tarifa.reset_index(drop=True, inplace=True)


CPU times: user 14.9 s, sys: 99.7 ms, total: 15 s
Wall time: 15 s


In [23]:
weatherStation_tarifa

DateTime  Tarifa_WindGust  Tarifa_Wind  Tarifa_WindDirection
0     2019-02-16T09:00:00        18.066667     9.066667                  72.0
1     2019-02-16T12:00:00        18.033333    11.233333                  86.0
2     2019-02-16T15:00:00        17.933333    10.566667                  82.0
3     2019-02-16T18:00:00        18.400000    12.000000                  87.0
4     2019-02-16T21:00:00              NaN          NaN                  83.0
...                   ...              ...          ...                   ...
6360  2021-05-21T18:00:00         4.433333     2.500000                 270.0
6361  2021-05-21T21:00:00         4.800000     2.466667                 275.0
6362  2021-05-22T00:00:00         5.333333     2.966667                 304.0
6363  2021-05-22T03:00:00         5.433333     2.900000                 297.0
6364  2021-05-22T06:00:00         5.050000     2.400000                 285.0

[6365 rows x 4 columns]

In [24]:
weatherStation_tarifa.isna().sum() # We still have a small percetenge of NaN to be treated later

DateTime                 0
Tarifa_WindGust         29
Tarifa_Wind             19
Tarifa_WindDirection    17
dtype: int64

- **STEP 2 WIND COMPONENTS** To allow Machine Learning models to treat the  wind direction, angles degree shall be descomposed into linear components N-S and E-W (the ML model wont be able to interpret that 359 º and 1 º are quite similar directions if we dont do this tranformation). 

In [25]:
# Function that return the linear component of an angular direction

def wind_components(wind_direction):
    NS_component = np.cos(np.radians(wind_direction))
    EW_component = np.sin(np.radians(wind_direction))
    return NS_component, EW_component

In [26]:
(weatherStation_tarifa['Tarifa_Direction N-S'],
 weatherStation_tarifa['Tarifa_Direction E-W']) = wind_components(weatherStation_tarifa['Tarifa_WindDirection'])

weatherStation_tarifa

DateTime  Tarifa_WindGust  Tarifa_Wind  Tarifa_WindDirection  \
0     2019-02-16T09:00:00        18.066667     9.066667                  72.0   
1     2019-02-16T12:00:00        18.033333    11.233333                  86.0   
2     2019-02-16T15:00:00        17.933333    10.566667                  82.0   
3     2019-02-16T18:00:00        18.400000    12.000000                  87.0   
4     2019-02-16T21:00:00              NaN          NaN                  83.0   
...                   ...              ...          ...                   ...   
6360  2021-05-21T18:00:00         4.433333     2.500000                 270.0   
6361  2021-05-21T21:00:00         4.800000     2.466667                 275.0   
6362  2021-05-22T00:00:00         5.333333     2.966667                 304.0   
6363  2021-05-22T03:00:00         5.433333     2.900000                 297.0   
6364  2021-05-22T06:00:00         5.050000     2.400000                 285.0   

      Tarifa_Direction N-S  Tarifa_Direction E-W  
0             3.090170e-01              0.951057  
1             6.975647e-02              0.997564  
2             1.391731e-01              0.990268  
3             5.233596e-02              0.998630  
4             1.218693e-01              0.992546  
...                    ...                   ...  
6360         -1.836970e-16             -1.000000  
6361          8.715574e-02             -0.996195  
6362          5.591929e-01             -0.829038  
6363          4.539905e-01             -0.891007  
6364          2.588190e-01             -0.965926  

[6365 rows x 6 columns]

In [27]:
weatherStation_tarifa

DateTime  Tarifa_WindGust  Tarifa_Wind  Tarifa_WindDirection  \
0     2019-02-16T09:00:00        18.066667     9.066667                  72.0   
1     2019-02-16T12:00:00        18.033333    11.233333                  86.0   
2     2019-02-16T15:00:00        17.933333    10.566667                  82.0   
3     2019-02-16T18:00:00        18.400000    12.000000                  87.0   
4     2019-02-16T21:00:00              NaN          NaN                  83.0   
...                   ...              ...          ...                   ...   
6360  2021-05-21T18:00:00         4.433333     2.500000                 270.0   
6361  2021-05-21T21:00:00         4.800000     2.466667                 275.0   
6362  2021-05-22T00:00:00         5.333333     2.966667                 304.0   
6363  2021-05-22T03:00:00         5.433333     2.900000                 297.0   
6364  2021-05-22T06:00:00         5.050000     2.400000                 285.0   

      Tarifa_Direction N-S  Tarifa_Direction E-W  
0             3.090170e-01              0.951057  
1             6.975647e-02              0.997564  
2             1.391731e-01              0.990268  
3             5.233596e-02              0.998630  
4             1.218693e-01              0.992546  
...                    ...                   ...  
6360         -1.836970e-16             -1.000000  
6361          8.715574e-02             -0.996195  
6362          5.591929e-01             -0.829038  
6363          4.539905e-01             -0.891007  
6364          2.588190e-01             -0.965926  

[6365 rows x 6 columns]

- **STEP 3 - WIND SPEED & WIND GUST IN KNOTS** as is the standard in nautical domains.

In [28]:
# Function that receive wind in m/s and return it in knots

def wind_knots(speed):
    return speed * 1.94384

In [29]:
weatherStation_tarifa['Tarifa_Wind'] = wind_knots(weatherStation_tarifa['Tarifa_Wind'])
weatherStation_tarifa['Tarifa_WindGust'] = wind_knots(weatherStation_tarifa['Tarifa_WindGust'])

weatherStation_tarifa

DateTime  Tarifa_WindGust  Tarifa_Wind  Tarifa_WindDirection  \
0     2019-02-16T09:00:00        35.118709    17.624149                  72.0   
1     2019-02-16T12:00:00        35.053915    21.835803                  86.0   
2     2019-02-16T15:00:00        34.859531    20.539909                  82.0   
3     2019-02-16T18:00:00        35.766656    23.326080                  87.0   
4     2019-02-16T21:00:00              NaN          NaN                  83.0   
...                   ...              ...          ...                   ...   
6360  2021-05-21T18:00:00         8.617691     4.859600                 270.0   
6361  2021-05-21T21:00:00         9.330432     4.794805                 275.0   
6362  2021-05-22T00:00:00        10.367147     5.766725                 304.0   
6363  2021-05-22T03:00:00        10.561531     5.637136                 297.0   
6364  2021-05-22T06:00:00         9.816392     4.665216                 285.0   

      Tarifa_Direction N-S  Tarifa_Direction E-W  
0             3.090170e-01              0.951057  
1             6.975647e-02              0.997564  
2             1.391731e-01              0.990268  
3             5.233596e-02              0.998630  
4             1.218693e-01              0.992546  
...                    ...                   ...  
6360         -1.836970e-16             -1.000000  
6361          8.715574e-02             -0.996195  
6362          5.591929e-01             -0.829038  
6363          4.539905e-01             -0.891007  
6364          2.588190e-01             -0.965926  

[6365 rows x 6 columns]

 ## 2.2 Historical Forecast DataFrame Preparation

For the purpose of this project and in order to simplify the data, we are going to maintain just one forecast for each "slice dt". The forecast selected will be the one made in the same day by 00.00 h that should be very accurate. 

***In other words, the forecast made by 00.00 hour will be kept as a prediction for the next 24 hours. The prediction interval for slice dt prediction for the first 24 hours is 3 hours, so we will maintain (24h/3h) 8 slice dt per forecast***

*Additionally* 
- We are going to kept just forecast from 6th of Feb 2019, that is the first record available at DataStations Datasets.
- Wind speed & gust in knots will be converted from m/s to knots
- Wind direction components N-S and E-W will be added

The strategy would be:

- **STEP 1** - Create a DataFrame with forecast dt from 2019-02-06 and on
- **STEP 2** - Filter forecast dt made by 00.00h
- **STEP 3** - From the previous DataFrame, keep just the first 8 slice dt (24h) from each forecast dt made by 00.00 hours
- **STEP 4** - Wind in knots
- **STEP 5** - Wind N-S & E-O components added

**STEP 1** - Create a dataset with forecast dt from 2019-02-06 and on

In [30]:
df_chunk = pd.read_csv('Data/historicalForecast.csv', chunksize=10000)

chunk_list = []

for chunk in df_chunk:
    chunk_list.append(chunk)

    
historicalForecast = pd.concat(chunk_list)

In [31]:
historicalForecast.shape

(591054, 23)

In [32]:
pd.set_option('display.max_rows', 10)

resultado = historicalForecast['forecast dt iso'].str.find('2019-02-06 00:00:00 +0000 UTC')
first_coincidence = list(resultado[resultado == 0].index)[0] 

In [33]:
# first coincidence in row 180420

historicalForecast = historicalForecast[historicalForecast.index>=first_coincidence]
historicalForecast.reset_index(drop=True, inplace=True)

In [34]:
historicalForecast

forecast dt unixtime                forecast dt iso  \
0                 1549411200  2019-02-06 00:00:00 +0000 UTC   
1                 1549411200  2019-02-06 00:00:00 +0000 UTC   
2                 1549411200  2019-02-06 00:00:00 +0000 UTC   
3                 1549411200  2019-02-06 00:00:00 +0000 UTC   
4                 1549411200  2019-02-06 00:00:00 +0000 UTC   
...                      ...                            ...   
410629            1621447200  2021-05-19 18:00:00 +0000 UTC   
410630            1621447200  2021-05-19 18:00:00 +0000 UTC   
410631            1621447200  2021-05-19 18:00:00 +0000 UTC   
410632            1621447200  2021-05-19 18:00:00 +0000 UTC   
410633            1621447200  2021-05-19 18:00:00 +0000 UTC   

        slice dt unixtime                   slice dt iso        lat       lon  \
0              1549411200  2019-02-06 00:00:00 +0000 UTC  36.013158 -5.598892   
1              1549422000  2019-02-06 03:00:00 +0000 UTC  36.013158 -5.598892   
2              1549432800  2019-02-06 06:00:00 +0000 UTC  36.013158 -5.598892   
3              1549443600  2019-02-06 09:00:00 +0000 UTC  36.013158 -5.598892   
4              1549454400  2019-02-06 12:00:00 +0000 UTC  36.013158 -5.598892   
...                   ...                            ...        ...       ...   
410629         1622786400  2021-06-04 06:00:00 +0000 UTC  36.013158 -5.598892   
410630         1622797200  2021-06-04 09:00:00 +0000 UTC  36.013158 -5.598892   
410631         1622808000  2021-06-04 12:00:00 +0000 UTC  36.013158 -5.598892   
410632         1622818800  2021-06-04 15:00:00 +0000 UTC  36.013158 -5.598892   
410633         1622829600  2021-06-04 18:00:00 +0000 UTC  36.013158 -5.598892   

        temperature  dew_point  pressure  ground_pressure  ...  wind_deg  \
0             13.04       8.45  1026.729         1024.826  ...    86.282   
1             12.76       7.48  1025.669         1023.900  ...    75.530   
2             12.64       7.45  1025.405         1023.779  ...    79.382   
3             12.99       8.35  1025.893         1024.346  ...    78.754   
4             14.66       8.65  1025.893         1024.496  ...    90.561   
...             ...        ...       ...              ...  ...       ...   
410629        17.80      15.58  1015.080         1002.681  ...   245.166   
410630        19.66      15.36  1015.997         1003.615  ...   257.380   
410631        20.85      15.54  1016.376         1004.008  ...   271.521   
410632        20.62      15.06  1015.688         1003.335  ...   267.430   
410633        19.08      15.31  1015.154         1002.786  ...   268.610   

        rain  snow  ice  fr_rain  convective             snow_depth  \
0        NaN   NaN  NaN      NaN         NaN  999900026055400900000   
1        0.0   0.0  0.0      0.0       0.000  999900026055400900000   
2        0.0   0.0  0.0      0.0       0.000  999900026055400900000   
3        0.0   0.0  0.0      0.0       0.000  999900026055400900000   
4        0.0   0.0  0.0      0.0       0.000  999900026055400900000   
...      ...   ...  ...      ...         ...                    ...   
410629   0.0   0.0  0.0      0.0      15.625                      0   
410630   0.0   0.0  0.0      0.0      15.625                      0   
410631   0.0   0.0  0.0      0.0      15.625                      0   
410632   0.0   0.0  0.0      0.0      15.625                      0   
410633   0.0   0.0  0.0      0.0      15.625                      0   

        accumulated  hours rate  
0               NaN    NaN  NaN  
1               0.0    3.0  0.0  
2               0.0    6.0  0.0  
3               0.0    3.0  0.0  
4               0.0    6.0  0.0  
...             ...    ...  ...  
410629          0.0    6.0  0.0  
410630          0.0    3.0  0.0  
410631          0.0    6.0  0.0  
410632          0.0    3.0  0.0  
410633          0.0    6.0  0.0  

[410634 rows x 23 columns]

**STEP 2** -Filter forecast dt made daily by 00.00 h

In [35]:
# 1 Filter forecast dt made by 00.00h

historicalForecast = historicalForecast[historicalForecast['forecast dt iso'].str.contains('00:00:00')]
historicalForecast.reset_index(drop=True, inplace=True)

**STEP 3** - Keep just the first 8 slice dt (24h) from each forecast dt made by 00.00 h

In [36]:
%%time
# 2 keep just the first 8 slice dt forecasts (24h) for each forecast dt by 00.00h

aux = historicalForecast.iloc[0:8,:]

for i in range (9,len(historicalForecast)):
    if historicalForecast['forecast dt iso'][i] == historicalForecast['slice dt iso'][i] :
        aux = pd.concat([aux,historicalForecast.iloc[i:i+8,:]])
        
historicalForecast = aux

CPU times: user 1.76 s, sys: 56.3 ms, total: 1.81 s
Wall time: 1.81 s


In [37]:
historicalForecast.reset_index(drop=True, inplace=True)
historicalForecast

forecast dt unixtime                forecast dt iso  slice dt unixtime  \
0               1549411200  2019-02-06 00:00:00 +0000 UTC         1549411200   
1               1549411200  2019-02-06 00:00:00 +0000 UTC         1549422000   
2               1549411200  2019-02-06 00:00:00 +0000 UTC         1549432800   
3               1549411200  2019-02-06 00:00:00 +0000 UTC         1549443600   
4               1549411200  2019-02-06 00:00:00 +0000 UTC         1549454400   
...                    ...                            ...                ...   
6667            1621382400  2021-05-19 00:00:00 +0000 UTC         1621414800   
6668            1621382400  2021-05-19 00:00:00 +0000 UTC         1621425600   
6669            1621382400  2021-05-19 00:00:00 +0000 UTC         1621436400   
6670            1621382400  2021-05-19 00:00:00 +0000 UTC         1621447200   
6671            1621382400  2021-05-19 00:00:00 +0000 UTC         1621458000   

                       slice dt iso        lat       lon  temperature  \
0     2019-02-06 00:00:00 +0000 UTC  36.013158 -5.598892        13.04   
1     2019-02-06 03:00:00 +0000 UTC  36.013158 -5.598892        12.76   
2     2019-02-06 06:00:00 +0000 UTC  36.013158 -5.598892        12.64   
3     2019-02-06 09:00:00 +0000 UTC  36.013158 -5.598892        12.99   
4     2019-02-06 12:00:00 +0000 UTC  36.013158 -5.598892        14.66   
...                             ...        ...       ...          ...   
6667  2021-05-19 09:00:00 +0000 UTC  36.013158 -5.598892        23.05   
6668  2021-05-19 12:00:00 +0000 UTC  36.013158 -5.598892        24.45   
6669  2021-05-19 15:00:00 +0000 UTC  36.013158 -5.598892        22.65   
6670  2021-05-19 18:00:00 +0000 UTC  36.013158 -5.598892        19.66   
6671  2021-05-19 21:00:00 +0000 UTC  36.013158 -5.598892        17.80   

      dew_point  pressure  ground_pressure  ...  wind_deg  rain  snow  ice  \
0          8.45  1026.729         1024.826  ...    86.282   NaN   NaN  NaN   
1          7.48  1025.669         1023.900  ...    75.530   0.0   0.0  0.0   
2          7.45  1025.405         1023.779  ...    79.382   0.0   0.0  0.0   
3          8.35  1025.893         1024.346  ...    78.754   0.0   0.0  0.0   
4          8.65  1025.893         1024.496  ...    90.561   0.0   0.0  0.0   
...         ...       ...              ...  ...       ...   ...   ...  ...   
6667      12.91  1018.467         1006.124  ...   231.962   0.0   0.0  0.0   
6668      11.55  1018.233         1005.913  ...    78.455   0.0   0.0  0.0   
6669      12.45  1018.215         1005.855  ...    68.753   0.0   0.0  0.0   
6670      16.55  1018.845         1006.444  ...    61.292   0.0   0.0  0.0   
6671      17.16  1020.865         1008.397  ...    59.161   0.0   0.0  0.0   

      fr_rain  convective             snow_depth  accumulated  hours rate  
0         NaN         NaN  999900026055400900000          NaN    NaN  NaN  
1         0.0         0.0  999900026055400900000          0.0    3.0  0.0  
2         0.0         0.0  999900026055400900000          0.0    6.0  0.0  
3         0.0         0.0  999900026055400900000          0.0    3.0  0.0  
4         0.0         0.0  999900026055400900000          0.0    6.0  0.0  
...       ...         ...                    ...          ...    ...  ...  
6667      0.0         0.0                      0          0.0    3.0  0.0  
6668      0.0         0.0                      0          0.0    6.0  0.0  
6669      0.0         0.0                      0          0.0    3.0  0.0  
6670      0.0         0.0                      0          0.0    6.0  0.0  
6671      0.0         0.0                      0          0.0    3.0  0.0  

[6672 rows x 23 columns]

- **STEP 4** - Wind in knots


In [38]:
historicalForecast['wind_speed'] = wind_knots(historicalForecast['wind_speed'])

- **STEP 5** - Wind N-S & E-W components added


In [39]:
(historicalForecast['Direction N-S'], historicalForecast['Direction E-W']) = wind_components(historicalForecast['wind_deg'])

In [40]:
pd.set_option('display.max_columns', None)
historicalForecast

forecast dt unixtime                forecast dt iso  slice dt unixtime  \
0               1549411200  2019-02-06 00:00:00 +0000 UTC         1549411200   
1               1549411200  2019-02-06 00:00:00 +0000 UTC         1549422000   
2               1549411200  2019-02-06 00:00:00 +0000 UTC         1549432800   
3               1549411200  2019-02-06 00:00:00 +0000 UTC         1549443600   
4               1549411200  2019-02-06 00:00:00 +0000 UTC         1549454400   
...                    ...                            ...                ...   
6667            1621382400  2021-05-19 00:00:00 +0000 UTC         1621414800   
6668            1621382400  2021-05-19 00:00:00 +0000 UTC         1621425600   
6669            1621382400  2021-05-19 00:00:00 +0000 UTC         1621436400   
6670            1621382400  2021-05-19 00:00:00 +0000 UTC         1621447200   
6671            1621382400  2021-05-19 00:00:00 +0000 UTC         1621458000   

                       slice dt iso        lat       lon  temperature  \
0     2019-02-06 00:00:00 +0000 UTC  36.013158 -5.598892        13.04   
1     2019-02-06 03:00:00 +0000 UTC  36.013158 -5.598892        12.76   
2     2019-02-06 06:00:00 +0000 UTC  36.013158 -5.598892        12.64   
3     2019-02-06 09:00:00 +0000 UTC  36.013158 -5.598892        12.99   
4     2019-02-06 12:00:00 +0000 UTC  36.013158 -5.598892        14.66   
...                             ...        ...       ...          ...   
6667  2021-05-19 09:00:00 +0000 UTC  36.013158 -5.598892        23.05   
6668  2021-05-19 12:00:00 +0000 UTC  36.013158 -5.598892        24.45   
6669  2021-05-19 15:00:00 +0000 UTC  36.013158 -5.598892        22.65   
6670  2021-05-19 18:00:00 +0000 UTC  36.013158 -5.598892        19.66   
6671  2021-05-19 21:00:00 +0000 UTC  36.013158 -5.598892        17.80   

      dew_point  pressure  ground_pressure  humidity  clouds  wind_speed  \
0          8.45  1026.729         1024.826    73.414     NaN    3.176235   
1          7.48  1025.669         1023.900    70.314     0.0    5.475797   
2          7.45  1025.405         1023.779    70.400     0.0    6.371908   
3          8.35  1025.893         1024.346    73.100     0.0    6.109489   
4          8.65  1025.893         1024.496    67.300     0.0    6.152254   
...         ...       ...              ...       ...     ...         ...   
6667      12.91  1018.467         1006.124    52.600    41.9    1.315980   
6668      11.55  1018.233         1005.913    44.200    28.7    4.089839   
6669      12.45  1018.215         1005.855    52.500    25.3   11.151810   
6670      16.55  1018.845         1006.444    82.600    21.5   11.254834   
6671      17.16  1020.865         1008.397    96.000    28.3   11.595006   

      wind_deg  rain  snow  ice  fr_rain  convective             snow_depth  \
0       86.282   NaN   NaN  NaN      NaN         NaN  999900026055400900000   
1       75.530   0.0   0.0  0.0      0.0         0.0  999900026055400900000   
2       79.382   0.0   0.0  0.0      0.0         0.0  999900026055400900000   
3       78.754   0.0   0.0  0.0      0.0         0.0  999900026055400900000   
4       90.561   0.0   0.0  0.0      0.0         0.0  999900026055400900000   
...        ...   ...   ...  ...      ...         ...                    ...   
6667   231.962   0.0   0.0  0.0      0.0         0.0                      0   
6668    78.455   0.0   0.0  0.0      0.0         0.0                      0   
6669    68.753   0.0   0.0  0.0      0.0         0.0                      0   
6670    61.292   0.0   0.0  0.0      0.0         0.0                      0   
6671    59.161   0.0   0.0  0.0      0.0         0.0                      0   

      accumulated  hours  rate  Direction N-S  Direction E-W  
0             NaN    NaN   NaN       0.064846       0.997895  
1             0.0    3.0   0.0       0.249873       0.968279  
2             0.0    6.0   0.0       0.184260       0.982878  
3             0.0    3.0   0.0       0.195022       0.

In [41]:
historicalForecast.columns

Index(['forecast dt unixtime', 'forecast dt iso', 'slice dt unixtime',
       'slice dt iso', 'lat', 'lon', 'temperature', 'dew_point', 'pressure',
       'ground_pressure', 'humidity', 'clouds', 'wind_speed', 'wind_deg',
       'rain', 'snow', 'ice', 'fr_rain', 'convective', 'snow_depth',
       'accumulated', 'hours', 'rate', 'Direction N-S', 'Direction E-W'],
      dtype='object')

In [42]:
historicalForecast.isna().sum() 

# we have quite a lot nan (833 out of 6672) in clouds column and in rain columns (accumalted and hours) to be trated latar

forecast dt unixtime      0
forecast dt iso           0
slice dt unixtime         0
slice dt iso              0
lat                       0
                       ... 
accumulated             833
hours                   833
rate                    131
Direction N-S             0
Direction E-W             0
Length: 25, dtype: int64

Documentation of the different fields

- main
 - main.pressure Atmospheric pressure (on the sea level), hPa
 - main.grnd_pressure Atmospheric pressure (on the surface), hPa
 - main.temp Temperature
 - main.dew_point Temperature of the dew point
 - main.humidity Humidity, %
-precipitation
 - precipitation.snow_depth Depth of snow, cm
 - precipitation.snow Snow water equivalent, mm
 - precipitation.rain Rain, mm
 - precipitation.hours Number of hours for which precipitation values are calculated
 - precipitation.rate Intensity, mm/s
 - precipitation.fr_rain Freezing rain, mm
 - precipitation.ice Ice pellets, mm
-wind
 - wind.speed Wind speed
 - wind.deg Wind direction, degrees (meteorological)
-clouds
 - clouds.all Cloudiness, %
 - dt Date for which the forecast is made, unix, UTC


# 3. Merging DataFrames
- **STEP 1** Addapt dates in the same format
- **STEP 2** merge dataframes with date as a key

**First, HistoricalForecast..**

**historicalForecast['slice dt iso'].pd.to_datetime()** does not work, we need to modicate the string before...



In [43]:
historicalForecast['slice dt iso'] = historicalForecast['slice dt iso'].str[:-10]
historicalForecast['forecast dt iso'] = historicalForecast['forecast dt iso'].str[:-10]

In [44]:
historicalForecast['slice dt iso'] = pd.to_datetime(historicalForecast['slice dt iso'])
historicalForecast['forecast dt iso'] = pd.to_datetime(historicalForecast['forecast dt iso'])

In [45]:
historicalForecast

forecast dt unixtime forecast dt iso  slice dt unixtime  \
0               1549411200      2019-02-06         1549411200   
1               1549411200      2019-02-06         1549422000   
2               1549411200      2019-02-06         1549432800   
3               1549411200      2019-02-06         1549443600   
4               1549411200      2019-02-06         1549454400   
...                    ...             ...                ...   
6667            1621382400      2021-05-19         1621414800   
6668            1621382400      2021-05-19         1621425600   
6669            1621382400      2021-05-19         1621436400   
6670            1621382400      2021-05-19         1621447200   
6671            1621382400      2021-05-19         1621458000   

            slice dt iso        lat       lon  temperature  dew_point  \
0    2019-02-06 00:00:00  36.013158 -5.598892        13.04       8.45   
1    2019-02-06 03:00:00  36.013158 -5.598892        12.76       7.48   
2    2019-02-06 06:00:00  36.013158 -5.598892        12.64       7.45   
3    2019-02-06 09:00:00  36.013158 -5.598892        12.99       8.35   
4    2019-02-06 12:00:00  36.013158 -5.598892        14.66       8.65   
...                  ...        ...       ...          ...        ...   
6667 2021-05-19 09:00:00  36.013158 -5.598892        23.05      12.91   
6668 2021-05-19 12:00:00  36.013158 -5.598892        24.45      11.55   
6669 2021-05-19 15:00:00  36.013158 -5.598892        22.65      12.45   
6670 2021-05-19 18:00:00  36.013158 -5.598892        19.66      16.55   
6671 2021-05-19 21:00:00  36.013158 -5.598892        17.80      17.16   

      pressure  ground_pressure  humidity  clouds  wind_speed  wind_deg  rain  \
0     1026.729         1024.826    73.414     NaN    3.176235    86.282   NaN   
1     1025.669         1023.900    70.314     0.0    5.475797    75.530   0.0   
2     1025.405         1023.779    70.400     0.0    6.371908    79.382   0.0   
3     1025.893         1024.346    73.100     0.0    6.109489    78.754   0.0   
4     1025.893         1024.496    67.300     0.0    6.152254    90.561   0.0   
...        ...              ...       ...     ...         ...       ...   ...   
6667  1018.467         1006.124    52.600    41.9    1.315980   231.962   0.0   
6668  1018.233         1005.913    44.200    28.7    4.089839    78.455   0.0   
6669  1018.215         1005.855    52.500    25.3   11.151810    68.753   0.0   
6670  1018.845         1006.444    82.600    21.5   11.254834    61.292   0.0   
6671  1020.865         1008.397    96.000    28.3   11.595006    59.161   0.0   

      snow  ice  fr_rain  convective             snow_depth  accumulated  \
0      NaN  NaN      NaN         NaN  999900026055400900000          NaN   
1      0.0  0.0      0.0         0.0  999900026055400900000          0.0   
2      0.0  0.0      0.0         0.0  999900026055400900000          0.0   
3      0.0  0.0      0.0         0.0  999900026055400900000          0.0   
4      0.0  0.0      0.0         0.0  999900026055400900000          0.0   
...    ...  ...      ...         ...                    ...          ...   
6667   0.0  0.0      0.0         0.0                      0          0.0   
6668   0.0  0.0      0.0         0.0                      0          0.0   
6669   0.0  0.0      0.0         0.0                      0          0.0   
6670   0.0  0.0      0.0         0.0                      0          0.0   
6671   0.0  0.0      0.0         0.0                      0          0.0   

      hours  rate  Direction N-S  Direction E-W  
0       NaN   NaN       0.064846       0.997895  
1       3.0   0.0       0.249873       0.968279  
2       6.0   0.0       0.184260       0.982878  
3       3.0   0.0       0.195022       0.980799  
4       6.0   0.0      -0.009791       0.999952  
...     ...   ...            ...            ...  
6667    3.0   0.0      -0.616184      -0.787602  
6668    6.0   0.0       0.200138       0.979768  
6669    3.0 

In [46]:
weatherStation_tarifa['DateTime'] = pd.to_datetime(weatherStation_tarifa['DateTime'])

In [47]:
weatherStation_tarifa

DateTime  Tarifa_WindGust  Tarifa_Wind  Tarifa_WindDirection  \
0    2019-02-16 09:00:00        35.118709    17.624149                  72.0   
1    2019-02-16 12:00:00        35.053915    21.835803                  86.0   
2    2019-02-16 15:00:00        34.859531    20.539909                  82.0   
3    2019-02-16 18:00:00        35.766656    23.326080                  87.0   
4    2019-02-16 21:00:00              NaN          NaN                  83.0   
...                  ...              ...          ...                   ...   
6360 2021-05-21 18:00:00         8.617691     4.859600                 270.0   
6361 2021-05-21 21:00:00         9.330432     4.794805                 275.0   
6362 2021-05-22 00:00:00        10.367147     5.766725                 304.0   
6363 2021-05-22 03:00:00        10.561531     5.637136                 297.0   
6364 2021-05-22 06:00:00         9.816392     4.665216                 285.0   

      Tarifa_Direction N-S  Tarifa_Direction E-W  
0             3.090170e-01              0.951057  
1             6.975647e-02              0.997564  
2             1.391731e-01              0.990268  
3             5.233596e-02              0.998630  
4             1.218693e-01              0.992546  
...                    ...                   ...  
6360         -1.836970e-16             -1.000000  
6361          8.715574e-02             -0.996195  
6362          5.591929e-01             -0.829038  
6363          4.539905e-01             -0.891007  
6364          2.588190e-01             -0.965926  

[6365 rows x 6 columns]

In [48]:
df = pd.merge(historicalForecast,weatherStation_tarifa, left_on='slice dt iso', right_on='DateTime', how ='inner')

In [49]:
df

forecast dt unixtime forecast dt iso  slice dt unixtime  \
0               1550275200      2019-02-16         1550307600   
1               1550275200      2019-02-16         1550318400   
2               1550275200      2019-02-16         1550329200   
3               1550275200      2019-02-16         1550340000   
4               1550275200      2019-02-16         1550350800   
...                    ...             ...                ...   
6341            1621382400      2021-05-19         1621414800   
6342            1621382400      2021-05-19         1621425600   
6343            1621382400      2021-05-19         1621436400   
6344            1621382400      2021-05-19         1621447200   
6345            1621382400      2021-05-19         1621458000   

            slice dt iso        lat       lon  temperature  dew_point  \
0    2019-02-16 09:00:00  36.013158 -5.598892        13.79      11.13   
1    2019-02-16 12:00:00  36.013158 -5.598892        13.81      11.16   
2    2019-02-16 15:00:00  36.013158 -5.598892        13.82      11.15   
3    2019-02-16 18:00:00  36.013158 -5.598892        13.77      11.27   
4    2019-02-16 21:00:00  36.013158 -5.598892        13.74      11.24   
...                  ...        ...       ...          ...        ...   
6341 2021-05-19 09:00:00  36.013158 -5.598892        23.05      12.91   
6342 2021-05-19 12:00:00  36.013158 -5.598892        24.45      11.55   
6343 2021-05-19 15:00:00  36.013158 -5.598892        22.65      12.45   
6344 2021-05-19 18:00:00  36.013158 -5.598892        19.66      16.55   
6345 2021-05-19 21:00:00  36.013158 -5.598892        17.80      17.16   

      pressure  ground_pressure  humidity  clouds  wind_speed  wind_deg  \
0     1021.601         1020.681    83.994   100.0   25.963871    95.202   
1     1022.066         1021.490    84.064    96.0   25.437090    93.856   
2     1020.887         1020.461    83.926    92.0   23.938390    93.883   
3     1020.964         1020.196    84.583    94.0   19.924360    95.812   
4     1022.636         1021.323    84.845    88.0   16.271885    94.213   
...        ...              ...       ...     ...         ...       ...   
6341  1018.467         1006.124    52.600    41.9    1.315980   231.962   
6342  1018.233         1005.913    44.200    28.7    4.089839    78.455   
6343  1018.215         1005.855    52.500    25.3   11.151810    68.753   
6344  1018.845         1006.444    82.600    21.5   11.254834    61.292   
6345  1020.865         1008.397    96.000    28.3   11.595006    59.161   

       rain  snow  ice  fr_rain  convective             snow_depth  \
0     0.125   0.0  0.0      0.0       0.125  999900026055400900000   
1     0.375   0.0  0.0      0.0       0.375  999900026055400900000   
2     0.188   0.0  0.0      0.0       0.188  999900026055400900000   
3     0.312   0.0  0.0      0.0       0.312  999900026055400900000   
4     0.000   0.0  0.0      0.0       0.000  999900026055400900000   
...     ...   ...  ...      ...         ...                    ...   
6341  0.000   0.0  0.0      0.0       0.000                      0   
6342  0.000   0.0  0.0      0.0       0.000                      0   
6343  0.000   0.0  0.0      0.0       0.000                      0   
6344  0.000   0.0  0.0      0.0       0.000                      0   
6345  0.000   0.0  0.0      0.0       0.000                      0   

      accumulated  hours  rate  Direction N-S  Direction E-W  \
0           0.125    3.0   0.0      -0.090667       0.995881   
1           0.375    6.0   0.0      -0.067249       0.997736   
2           0.188    3.0   0.0      -0.067719       0.997704   
3           0.312    6.0   0.0      -0.101265       0.994860   
4           0.000    3.0   0.0      -0.073464       0.997298   
...           ...    ...   ...            ...            ...   
6341        0.000    3.0   0.0      -0.616184      -0.787602   
6342        0.000    6.0   0.0       0.200138       0.979768   
6343        0.000    3.0   0.0

In [50]:
pd.set_option('display.max_rows', 33) 
df.isnull().sum()

forecast dt unixtime      0
forecast dt iso           0
slice dt unixtime         0
slice dt iso              0
lat                       0
lon                       0
temperature               0
dew_point                 0
pressure                  0
ground_pressure           0
humidity                  0
clouds                  793
wind_speed                0
wind_deg                  0
rain                    120
snow                    120
ice                     120
fr_rain                 120
convective              793
snow_depth                0
accumulated             793
hours                   793
rate                    120
Direction N-S             0
Direction E-W             0
DateTime                  0
Tarifa_WindGust          29
Tarifa_Wind              19
Tarifa_WindDirection     17
Tarifa_Direction N-S     17
Tarifa_Direction E-W     17
dtype: int64

# 4. Exporting to a csv file

In [51]:
df.to_csv('Data/windProject.csv', index=False)